### PR TITLE
feat(query): Refactor query engine to allow for new features

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Refactored Query Engine into separate classes to allow for new Query Engine features to be introduced.

* Using decorator pattern on `QueryPlanner` for multi-partition, multi-cluster, downsampling features. Moved different features into different planners.
* Engine works by composing different types of `QueryPlanner` objects to provide overall capability
* Folded TsdbQueryParam, queryId, submitTime into QueryOptions to reduce number of params in various methods. Renamed QueryOptions to QueryContext
* Renamed RoutingPlanner to RoutingStrategy
* Rename QueryEngine to CompositePlanner
* Rename package queryengine2 to queryplanner
* Split QueryEngineSpec into multiple planner specs

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala [hunk 1]
```diff
@@ -39,595 +23,32 @@ object UnavailablePromQlQueryParams extends TsdbQueryParams
...
+  //val multiPodPlanner = new MultiPodPlanner(podLocalityProvider, haPlanner)
 
   /**
     * This is the facade to trigger orchestration of the ExecPlan.
     * It sends the ExecPlan to the destination where it will be executed.
     */
-  def dispatchExecPlan(execPlan: ExecPlan)
-                      (implicit sched: Scheduler,
-                       timeout: FiniteDuration): Task[QueryResponse] = {
+  def dispatchExecPlan(execPlan: ExecPlan)(implicit sched: Scheduler, timeout: FiniteDuration): Task[QueryResponse] = {
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Perhaps this utility can go into QueryPlanner? Then QueryEngine goes away. It's really a way of gluing these planners together.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala [hunk 2]
```diff
@@ -39,595 +23,32 @@ object UnavailablePromQlQueryParams extends TsdbQueryParams
...
-    */
-  private case class PlanResult(plans: Seq[ExecPlan], needsStitch: Boolean = false)
+                  queryEngineConfig: Config = ConfigFactory.empty()) extends QueryPlanner with StrictLogging {
 
-  private val mdNoShardKeyFilterRequests = Kamon.counter("queryengine-metadata-no-shardkey-requests")
-  private val dsOptions = schemas.part.options
+  // Note the composition of query planners below using decorator pattern
+  val rawClusterPlanner = new SingleClusterPlanner(dsRef, schemas, spreadProvider, shardMapperFunc)
+  val downsampleClusterPlanner = new SingleClusterPlanner(dsRef, schemas, spreadProvider, shardMapperFunc)
+  val downsampleStitchPlanner = new DownsampleStitchPlanner(rawClusterPlanner, downsampleClusterPlanner)
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Is this intended to be a generic "stitcher"? If so, don't call it "downsample" stitcher.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/MultiPodPlanner.scala [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package filodb.coordinator.queryengine2
+
+import filodb.query.{LogicalPlan, QueryContext}
+import filodb.query.exec.ExecPlan
+
+/**
+  * MultiPodPlanner is responsible for breaking down queries into separate subqueries,
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/queryengine2/MultiPodPlanner.scala]: Perhaps remove this for now and move the comments to where you intend to use this.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/HighAvailabilityPlanner.scala [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        else
+          localPlanner.materialize(
+            copyWithUpdatedTimeRange(rootLogicalPlan, route.asInstanceOf[LocalRoute].timeRange.get, lookBackTime),
+            options)
+        case route: RemoteRoute =>
+          val timeRange = route.timeRange.get
+          val queryParams = options.origQueryParams.asInstanceOf[PromQlQueryParams]
+          val routingConfig = queryEngineConfig.getConfig("routing")
+          val promQlInvocationParams = PromQlInvocationParams(routingConfig, queryParams.promQl,
+            timeRange.startInMillis / 1000, queryParams.step, timeRange.endInMillis / 1000,
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/queryengine2/HighAvailabilityPlanner.scala]: This division by 1000 really confuses me. Some comments will help.

File: coordinator/src/main/scala/filodb.coordinator/QueryActor.scala [hunk 5]
```diff
@@ -119,15 +119,11 @@ final class QueryActor(memStore: MemStore,
      }(queryScheduler)
   }
 
-  private def getSpreadProvider(queryOptions: QueryOptions): SpreadProvider = {
-    return queryOptions.spreadProvider.getOrElse(functionalSpreadProvider)
-  }
-
   private def processLogicalPlan2Query(q: LogicalPlan2Query, replyTo: ActorRef) = {
-    // This is for CLI use only. Always prefer clients to materialize logical plan
+    // This is for CLI use only. Always prefer clients to materializeHaPlan logical plan
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/QueryActor.scala]: materializeHaPlan?

File: coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala [hunk 6]
```diff
@@ -137,7 +137,7 @@ class QueryEngineSpec extends FunSpec with Matchers {
     }
   }
 
-  it("should materialize ExecPlan correctly for _bucket_ histogram queries") {
+  it("should materializeHaPlan ExecPlan correctly for _bucket_ histogram queries") {
```
Comment (broneill) [inline coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala]: materializeHaPlan?

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala (hunk 1 above)
Comment (vishramachandran) [reply coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: Good idea.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/HighAvailabilityPlanner.scala (hunk 4 above)
Comment (vishramachandran) [reply coordinator/src/main/scala/filodb.coordinator/queryengine2/HighAvailabilityPlanner.scala]: PromQL params are in seconds. Added comment.

File: coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala (hunk 6 above)
Comment (vishramachandran) [reply coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala]: Bad intellij rename. Reverted.

File: coordinator/src/main/scala/filodb.coordinator/QueryActor.scala (hunk 5 above)
Comment (vishramachandran) [reply coordinator/src/main/scala/filodb.coordinator/QueryActor.scala]: Bad intellij rename. Reverted.

File: coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala (hunk 2 above)
Comment (vishramachandran) [reply coordinator/src/main/scala/filodb.coordinator/queryengine2/QueryEngine.scala]: I am renaming class to LongTimeRangePlanner

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/CompositePlanner.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+import filodb.coordinator.ShardMapper
+import filodb.coordinator.client.QueryCommands.StaticSpreadProvider
+import filodb.core.{DatasetRef, SpreadProvider}
+import filodb.core.metadata.Schemas
+import filodb.query._
+import filodb.query.exec._
+
+/**
+  * FiloDB Query Engine is the facade for execution of FiloDB queries.
```
Comment (broneill) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/CompositePlanner.scala]: Comments now stale.

**Pull request merged**